### ISSUE TITLE
Allow to use CSS-Functions for Dimension properties

### ISSUE THREAD

**Issue by tomsontom** [adobe/react-spectrum#1716]
# 🙋 Feature Request

Currently one can use %,em, ... but it is not possible to use CSS functions like min/max/calc/clamp to size components using CSS-Functions

## 🤔 Expected Behavior

Pass through min/max/calc/clamp as it is currently done for %,em, ...

## 😯 Current Behavior

Value is not passed

## 💁 Possible Solution

Add a check for min/max/calc/clamp - if you agree this is a good idea I'd prepare a PR.

## 🔦 Context

I'd like to size elements using some clever CSS-Functions

**Comment by snowystinger** [comment]
I imagine this will have the same issues as https://github.com/adobe/react-spectrum/issues/1140
If you have any API suggestions, please add them here.

**Comment by tomsontom** [comment]
well it is not 100% the same - i don't really ask for support of that at the API-Level nor (at least at first) to honor spectrum-variables there. On the API-Level `DimensionValue` already allows today to type in an arbitary string. 

My current request is that `dimensionValue()`-function in `styleProps.ts` recognizes those functions like it today already does recognize %,... . The method would change to something like this (totally untested)

```typescript
const UNIT_RE = /(%|px|em|rem|vw|vh|auto|cm|mm|in|pt|pc|ex|ch|rem|vmin|vmax|fr)$/;
const FUNC_RE = /^calc|clamp|max|min/;
export function dimensionValue(value: DimensionValue) {
  if (typeof value === 'number') {
    return value + 'px';
  }

  if (UNIT_RE.test(value) || FUNC_RE.test(value)) {
    return value;
  }

  return `var(--spectrum-global-dimension-${value}, var(--spectrum-alias-${value}))`;
}
```

**Comment by snowystinger** [comment]
so some examples of this in action might be

```jsx
<View width="size-300" />
<View width="20%" />
// and newly added
<View width="calc(20% + 15px)" />
// but would someone also expect they could do this
<View width="calc(20% + size-300" />
```
that last one wouldn't work, but seems totally valid to me
should we try to account for that last one by matching "size-[0-9]*?"
there's also `single-line-width` and `single-line-height` https://react-spectrum.adobe.com/react-spectrum/styling.html#dimension-values

**Comment by tomsontom** [comment]
Right we could get as sophisticated as you want and I'd work on that if you all agree that's something you consider to add

**Comment by snowystinger** [comment]
We'd be happy to explore this with you. Would you mind putting together a WIP or detailing out more here?
Our feeling is that these CSS functions would not be the most useful if they didn't allow for the spectrum variables, so we'd like to include that from the get go.

**Comment by tomsontom** [comment]
Well there's a very simple solution but to make it generic we'd have to change the definition of DimensionValue using some typescript typesystem trick (IIRC it is available since 3.4 or something like that).

```typescript
const SpectrumDimensionValues = ['size-0'
  , 'size-10'
  , 'size-25'
  , 'size-40'
  , 'size-50'
  , 'size-65'
  , 'size-75'
  , 'size-85'
  , 'size-100'
  , 'size-115'
  , 'size-125'
  , 'size-130'
  , 'size-150'
  , 'size-160'
  , 'size-175'
  , 'size-200'
  , 'size-225'
  , 'size-250'
  , 'size-300'
  , 'size-350'
  , 'size-400'
  , 'size-450'
  , 'size-500'
  , 'size-550'
  , 'size-600'
  , 'size-675'
  , 'size-700'
  , 'size-800'
  , 'size-900'
  , 'size-1000'
  , 'size-1200'
  , 'size-1250'
  , 'size-1600'
  , 'size-1700'
  , 'size-2000'
  , 'size-2400'
  , 'size-3000'
  , 'size-3400'
  , 'size-3600'
  , 'size-4600'
  , 'size-5000'
  , 'size-6000'
  , 'static-size-0'
  , 'static-size-10'
  , 'static-size-25'
  , 'static-size-50'
  , 'static-size-40'
  , 'static-size-65'
  , 'static-size-100'
  , 'static-size-115'
  , 'static-size-125'
  , 'static-size-130'
  , 'static-size-150'
  , 'static-size-160'
  , 'static-size-175'
  , 'static-size-200'
  , 'static-size-225'
  , 'static-size-250'
  , 'static-size-300'
  , 'static-size-400'
  , 'static-size-450'
  , 'static-size-500'
  , 'static-size-550'
  , 'static-size-600'
  , 'static-size-700'
  , 'static-size-800'
  , 'static-size-900'
  , 'static-size-1000'
  , 'static-size-1200'
  , 'static-size-1700'
  , 'static-size-2400'
  , 'static-size-2600'
  , 'static-size-3400'
  , 'static-size-3600'
  , 'static-size-4600'
  , 'static-size-5000'
  , 'static-size-6000'
  , 'single-line-height'
  , 'single-line-width'] as const;

type DimensionValue =
  typeof SpectrumDimensionValues[number]
  | (string & {})
  | number;

const SIMPLE_REG = new RegExp(SpectrumDimensionValues.join('|'),'g');

console.log('min(clamp(calc((static-size-1200 + 10px)*1.5px), 2.5vw, 2.8rem), 20px)'
  .replace(SIMPLE_REG,'var(--spectrum-global-dimension-$&, var(--spectrum-alias-$&))')); 
// prints min(clamp(calc((var(--spectrum-global-dimension-static-size-1200, var(--spectrum-alias-static-size-1200)) + 10px)*1.5px), 2.5vw, 2.8rem), 20px)
```

I've also written a version where I tried to parse the functions but the problem is that `-` is used inside the spectrum constants but that is also a valid arithmetic operator inside all this functions eg `10px-5px` so I'd have to tune my algorithm but IMHO the simplicity of the RegExp version is too obvious.

**Comment by tomsontom** [comment]
Argh I see "DimensionValue" is defined in a `d.ts` so the trick does not work, I need to see if I can get a generic version can work somehow
